common/common_test.go

```diff
@@ -0,0 +1,290 @@
+package common
+
+import (
+	"bufio"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/zcash/lightwalletd/walletrpc"
+)
+
+// ------------------------------------------ Setup
+//
+// This section does some setup things that may (even if not currently)
+// be useful across multiple tests.
+
+var (
+	testT *testing.T
+
+	// The various stub callbacks need to sequence through states
+	step int
+
+	getblockchaininfoReply []byte
+	logger                 = logrus.New()
+
+	getsaplinginfo []byte
+
+	blocks [][]byte // four test blocks
+)
+
+// TestMain does common setup that's shared across multiple tests
+func TestMain(m *testing.M) {
+	output, err := os.OpenFile("test-log", os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0644)
+	if err != nil {
+		os.Stderr.WriteString(fmt.Sprintf("Cannot open test-log: %v", err))
+		os.Exit(1)
+	}
+	logger.SetOutput(output)
+	Log = logger.WithFields(logrus.Fields{
+		"app": "test",
+	})
+
+	getsaplinginfo, err := ioutil.ReadFile("../testdata/getsaplinginfo")
+	if err != nil {
+		os.Stderr.WriteString(fmt.Sprintf("Cannot open testdata/getsaplinginfo: %v", err))
+		os.Exit(1)
+	}
+	getblockchaininfoReply, _ = hex.DecodeString(string(getsaplinginfo))
+
+	// Several tests need test blocks; read all 4 into memory just once
+	// (for efficiency).
+	testBlocks, err := os.Open("../testdata/blocks")
+	if err != nil {
+		os.Stderr.WriteString(fmt.Sprintf("Cannot open testdata/blocks: %v", err))
+		os.Exit(1)
+	}
+	scan := bufio.NewScanner(testBlocks)
+	for scan.Scan() { // each line (block)
+		block := scan.Bytes()
+		// Enclose the hex string in quotes (to make it json, to match what's
+		// returned by the RPC)
+		block = []byte("\"" + string(block) + "\"")
+		blocks = append(blocks, block)
+	}
+
+	// Setup is done; run all tests.
+	exitcode := m.Run()
+
+	// cleanup
+	os.Remove("test-log")
+
+	os.Exit(exitcode)
+}
+
+// Allow tests to verify that sleep has been called (for retries)
+var sleepCount int
+var sleepDuration time.Duration
+
+func sleepStub(d time.Duration) {
+	sleepCount++
+	sleepDuration += d
+}
+
+// ------------------------------------------ GetSaplingInfo()
+
+func getblockchaininfoStub(method string, params []json.RawMessage) (json.RawMessage, error) {
+	step++
+	// Test retry logic (for the moment, it's very simple, just one retry).
+	switch step {
+	case 1:
+		return getblockchaininfoReply, errors.New("first failure")
+	}
+	if sleepCount != 1 || sleepDuration != 15*time.Second {
+		testT.Error("unexpected sleeps", sleepCount, sleepDuration)
+	}
```

Comment:
It may be overkill to verify the sleep times... We can always remove this checking later.